net/socket.h

```diff
@@ -0,0 +1,259 @@
+     */
+    void read(void *buffer, size_t len)
+    {
+        std::lock_guard<std::mutex> guard(m_ReadMutex);
+        checkSocket();
+
+        // initially, copy over any leftover bytes in m_Buffer
+        size_t start = 0;
+        if (m_BufferBytes > 0) {
+            size_t tocopy = std::min(len, m_BufferBytes);
+            memcpy(buffer, &m_Buffer[m_BufferStart], tocopy);
+            start += tocopy;
+            len -= tocopy;
+            debitBytes(tocopy);
+        }
+
+        // keep reading from socket until we have enough bytes
+        while (len > 0) {
+            size_t nbytes = readOnce((char *) buffer, start, len);
+            start += nbytes;
+            len -= nbytes;
+        }
+    }
+
+    /*
+     * Read a single line in, stopping at a newline char.
+     */
+    std::string readLine()
+    {
+        std::lock_guard<std::mutex> guard(m_ReadMutex);
+        checkSocket();
+
+        std::ostringstream oss;
+        while (true) {
+            if (m_BufferBytes == 0) {
+                m_BufferBytes += readOnce(m_Buffer,
+                                          m_BufferStart,
+                                          DefaultBufferSize - m_BufferStart);
+            }
+
+            // look for newline char
+            for (size_t i = 0; i < m_BufferBytes; i++) {
+                char &c = m_Buffer[m_BufferStart + i];
+                if (c == '\n') {
+                    c = '\0';
+                    oss << std::string(&m_Buffer[m_BufferStart]);
+                    debitBytes(i + 1);
+
+                    std::string outstring = oss.str();
+                    if (m_Print) {
+                        std::cout << "<<< " << outstring << std::endl;
+                    }
+                    return outstring;
+                }
+            }
+
+            // if newline is not present, append the text we received and try
+            // another read
+            oss << std::string(&m_Buffer[m_BufferStart], m_BufferBytes);
+            debitBytes(m_BufferBytes);
+        }
+    }
+
+    /*
+     * Send a buffer of specified length through the socket.
+     */
+    void send(const void *buffer, size_t len)
+    {
+        std::lock_guard<std::mutex> guard(m_SendMutex);
+        checkSocket();
+
+        int ret = ::send(m_Socket, (sendbuff_t) buffer, (bufflen_t) len, MSG_NOSIGNAL);
+        if (ret == -1) {
+            throw socket_error("Could not send " + errorMessage());
+        }
+    }
+
+    /*
+     * Send a string over the socket.
+     */
+    void send(const std::string &msg)
+    {
+        send(msg.c_str(), msg.length());
+
+        if (m_Print) {
+            std::cout << ">>> " << msg;
+        }
+    }
+
+    /*
+     * Set the current socket;
+     */
+    void setSocket(socket_t sock)
+    {
+        std::lock_guard<std::mutex> guard(m_ReadMutex);
+
+        m_Socket = sock;
+        checkSocket();
+    }
+
+private:
+    char m_Buffer[DefaultBufferSize];
```

Comment:
Windows particularly has a very small stack size and fails in a singularly difficult to diagnose way when it overflows so so I think it would be better to use a `std::vector ` and resize if to DefaultBufferSize in the contructor

Reply:
Good shout